working/unquoted-imports/feature-specification.md

```diff
@@ -281,7 +315,37 @@ import strange /* comment */    .   but
       fine;
 ```
 
-But they can also choose to *not* do that.
+This wouldn't cause any problems for a Dart implementation. It would simply
+discard the whitespace and comments as it does elsewhere and the resulting path
+is `strange.but/another/fine`.
+
+However, it likely causes problems for Dart *users* and other simpler tools and
+scripts that work with Dart code. In particular, we often see homegrown tools
+that want to "parse" a Dart file to find its package references and traverse the
+dependency graph. While these tools ideally should use a full Dart parser (like
+the one in the [analyzer package][], which is freely available), the reality is
+that users often cobble together simple scripts using regex to do this kind of
+parsing, or they need to write these tools in a language other than Dart. In
+those cases, if the package path happens to contain whitespace or comments, the
+tool will likely silently fail to recognize the package path.
+
+[analyzer package]: https://pub.dev/packages/analyzer
+
+Also, we find no compelling *use* for whitespace and comments inside package
+paths. To that end, this proposal makes it an error. All of the tokens in the
+path must be directly adjacent with no whitespace, newlines, or comments between
+them. The previous import is an error. However, we still allow comments in or
+after the directives outside of the path. These are all valid:
+
+```dart
+import /* Weird but OK. */ some/path;
+export some/path; // Hi there.
+part some/path // Before the semicolon? Really?
```

Comment:
This should probably be filed as an issue so that we don't lose the discussion. :)

For now, I'll leave the proposal as-is, but I'm definitely keen to talk about this more. For what it's worth, I would ultimately like to support unquoted parts. It's true that we don't support relative unquoted references. But my guess is that users will end up preferring unquoted package references even to files in the same package versus quoted relative imports. At least, that's what I think *I'll* end up doing.

If that's the case, then unquoted part directives will make sense too.

Reply:
> preferring unquoted package references even to files in the same package versus quoted relative imports

I find it unlikely that most people will prefers `part mypkg/src/subfeature/featurepart;` to `part "featurepart.dart";` in `mypkg/src/subfeature.feature.dart`.

We don't even have a way to omit the package name, and parts will always at least be below `src/` too, and almost always in the same directory as the library file.

It also only works for libraries inside `lib/`. If you have part files in `test/`, `bin/`, `tool/` or (now) `hook/`, you can't use unquoted package URIs for those.

But we'll see :Grin:. (I expect users will ask for ways to not quote relative paths, and we *can* give them that, as `part ./featurepart;`. I still think we should use `:` to separate package name from path, because that makes it harder to typo a relative path into a package path. But the most likely mistake is to do `import foo;` meaning `import "./foo.dart";`, which doesn't have any separator, so the separator probably won't make a big difference.)